src/app/app.component.ts

```diff
@@ -1,10 +1,78 @@
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormBuilder, FormControl, FormArray, Validators, ValidatorFn, AbstractControl } from '@angular/forms';
+import { HeroService } from './test-service/hero.service';
+import { Hero, HeroPower } from './test-service/hero.interface';
+import { Utils } from './utils';
 
 @Component({
-  selector: 'app-root',
-  templateUrl: './app.component.html',
-  styleUrls: ['./app.component.css']
+    selector: 'app-root',
+    templateUrl: './app.component.html',
+    styleUrls: ['./app.component.css']
 })
-export class AppComponent {
-  title = 'reactive-forms-example';
+export class AppComponent implements OnInit {
+    loading = true;
+    results = 'No results yet, click on save';
+
+    form: FormGroup;
+
+    static createPowerControl(name: string, description: string): FormGroup {
+        return new FormGroup({
+            name: new FormControl(name, Validators.required),
+            description: new FormControl(description),
+        });
+    }
+
+    constructor (
+        private heroService: HeroService,
+    ) { }
+
+    ngOnInit() {
+        this.heroService.getAsyncData()
+            .subscribe(hero => {
+                this.initForm(hero);
+                this.loading = false;
+            });
+    }
+
+    initForm(hero: Hero) {
+        const powerControls = hero.powers.map(power => AppComponent.createPowerControl(power.name, power.description));
+
+        this.form = new FormGroup({
+            name: new FormGroup({
+                alias: new FormControl(hero.name.alias, Validators.required),
+                realName: new FormControl(hero.name.realName),
+            }),
+            appearances: new FormGroup({
+                comics: new FormControl(hero.appearances.comics),
+                movies: new FormControl(hero.appearances.movies),
+                games: new FormControl(hero.appearances.games),
+            }),
+            powers: new FormArray(powerControls, this.getPowersValidator()),
+        });
+    }
+
+    getPowersValidator(): ValidatorFn {
+        return (control: FormArray) => {
+            if (control.length < 1) {
+                return { powersValidation: 'tooLittle' };
+            }
+            if (control.length > 5) {
+                return { powersValidation: 'tooMuch' };
+            }
+            return null;
+        };
+    }
+
+    save() {
+        if (this.form.valid) {
+            this.results = 'Saving data:<br /><br />' + JSON.stringify(this.form.value, null, 4);
+        } else {
+            Utils.validateFormControls(this.form.controls);
```

Comment:
Why are `markAsTouched()` and `markAsUntouched()` behaving inconsistently?  `Untouched` marks both children and parents as untouched while `Touched` only marks the parents. This means that I need an utility method to go through the whole form and mark every field manually as touched to highlight the validation (which only happens in my case on `.invalid.touched`)